prereise/gather/griddata/hifld/data_process/transmission.py

```diff
@@ -322,6 +328,105 @@ def create_transformers(bus):
+        reactance_lookup = pd.Series(const.line_reactance_per_mile)
+        # Find the closest voltage
+        closest_voltage = (reactance_lookup - branch.loc["VOLTAGE"]).abs().idxmin()
+        reactance_per_mile = reactance_lookup.loc[closest_voltage]
+        return branch.loc["length"] * reactance_per_mile
+    else:
+        raise ValueError(f"{branch.loc['type']} not a valid branch type")
+
+
+def calculate_branch_mileage(branch):
+    """Estimate distance of a line.
+
+    :param pandas.Series branch: data for a single line.
+    :return: (*float*) -- distance (in miles) for that line.
+    """
+    coordinates = branch.loc["COORDINATES"]
+    return sum([haversine(a, b) for a, b in zip(coordinates[:-1], coordinates[1:])])
+
+
+def estimate_branch_rating(branch):
+    """Estimate branch rating using line voltage or constant value for transformers.
+
+    :param pandas.Series branch: data for a single branch (line or transformer). All
+        branches require 'type' attributes, lines require 'VOLTAGE' and 'length'.
+    """
+    if branch.loc["type"] == "Line":
+        if branch.loc["length"] <= const.line_rating_short_threshold:
+            rating_lookup = pd.Series(const.line_rating_short)
+            closest_voltage = (rating_lookup - branch.loc["VOLTAGE"]).abs().idxmin()
+            return rating_lookup.loc[closest_voltage]
+        else:
+            rating_lookup = pd.Series(const.line_rating_surge_impedance_loading)
+            closest_voltage = (rating_lookup - branch.loc["VOLTAGE"]).abs().idxmin()
+            sil = rating_lookup.loc[closest_voltage]
+            return (
+                sil
+                * const.line_rating_surge_impedance_coefficient
+                * branch.loc["VOLTAGE"] ** const.line_rating_surge_impedance_exponent
+            )
+    elif branch.loc["type"] == "Transformer":
+        return const.transformer_rating
+    else:
+        raise ValueError(f"{branch.loc['type']} not a valid branch type")
+
+
+def get_mst_edges(lines, substations):
+    """Get the set of lines which connected the connected components of the lines graph,
+    either from a cache or by generating from scratch and caching.
+
+    :param pandas.DataFrame lines: data frame of lines.
+    :param pandas.DataFrame substations: data frame of substations.
+    :return: (*list*) -- each entry is a 3-tuple:
+        index of the first connected component (int),
+        index of the second connected component (int),
+        a dictionary with keys containing ``start``, ``end`` and ``weight``, defines
+        the ``from substation ID``, ``to substation ID`` and the distance of the line.
+    """
+    cache_dir = os.path.join(os.path.dirname(__file__), "cache")
+    os.makedirs(cache_dir, exist_ok=True)
+    cache_key = repr(lines[["SUB_1_ID", "SUB_2_ID"]].to_numpy().tolist()) + repr(
+        substations[["LATITUDE", "LONGITUDE"]].to_numpy().tolist()
+    )
```

Comment:
Given we are going to encode it anyways, do we need `to_numpy()` and `tolist()` here?

Reply:
Calling `repr` on a data frame or numpy array seems to truncate for large arrays:
```python
>>> import numpy as np
>>> a = np.arange(100)
>>> df = pd.DataFrame({"a": a})
>>> print(repr(df))
     a
0    0
1    1
2    2
3    3
4    4
..  ..
95  95
96  96
97  97
98  98
99  99

[100 rows x 1 columns]
```

numpy will happily display 100-length arrays, even 1,000-length arrays, but starts to truncate output somewhere between there and 10,000:
```python
>>> a = np.arange(10000)
>>> print(repr(a))
array([   0,    1,    2, ..., 9997, 9998, 9999])
```

Reply:
Does it imply it will give us same cache_key for different big data frames or it is simply an issue for display?

Reply:
Yes.
```python
>>> import numpy as np
>>> import pandas as pd
>>> a = np.arange(100)
>>> df1 = pd.DataFrame({"a": a})
>>> df2 = pd.DataFrame({"a": a})
>>> df2.loc[5:94] = -999
>>> df2
     a
0    0
1    1
2    2
3    3
4    4
..  ..
95  95
96  96
97  97
98  98
99  99

[100 rows x 1 columns]
>>> repr(df1) == repr(df2)
True
>>> df1.equals(df2)
False
```